[Macros] Add support for wasm macros

kabiroberai opened the pull request:
<!-- What's in this pull request? -->
This PR introduces support for WebAssembly-based macros via `-load-plugin-executable Foo.wasm#Foo`.

Wasm macros are blazing fast (no SwiftSyntax build required for consumers!) and secure (fully sandboxed on all OSes)
```
time $BIN/swiftc Client.swift \
  -load-plugin-executable ExampleRaw.wasm#ExampleRaw
Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swiftc -c 0.06s user 0.03s system 33% cpu 0.244 total
```

The approach I've taken here is to update `swift-plugin-server` to add special handling for Wasm macros. These macros are executed using WasmKit, but in the future we can [add support](https://github.com/kabiroberai/swift/compare/kabir/wasm-plugins..kabir/jsc-wasm) for other (faster, JIT) runtimes like JavaScriptCore.

Note also that these wasm macros currently have an ad-hoc ABI as I wanted to avoid a hard dependency on WASI. It's thus theoretically possible to build macros without a dependency on WASI, though I've had trouble doing so because Swift currently builds for wasm as WASI xor Embedded Mode — we'd need a middle ground with support for String etc. The current ABI is based on the one I built for [Wacro](https://github.com/kabiroberai/Wacro/blob/6416cd7d3904bce08215ffc044c0d1be9f666df3/Sources/WacroPluginRaw/Exports.swift). We'll have to iron out the execution model though: my current mental model is that plugins are request-response style (i.e. every HostToPluginMessage must be followed by exactly one PluginToHostMessage) and so the macro effectively exports a `char *macro_parse(char *request_json)` but I don't know if this is a guaranteed contract.

There's currently a few things that still need to be done:
- [x] Support other platforms with WasmKit. (*)
- [x] Unit tests

Followups:
- Figure out the story for building macro targets — the approach I'm using right now is to have the "source" macro depend on [Wacro](https://github.com/kabiroberai/Wacro) but we'll probably want to upstream some sort of Wasm macro module into swift-syntax. (*)
- SwiftPM support + evolution proposal? I'm envisioning a `.binaryMacro` target similar to `.binaryTarget`, though this will make the edit-compile-run cycle non-trivial as it splits building the macro into a separate phase. (*)

Marking this as a draft until the above items are done but I wanted to make this PR so I could get some input as well as to brainstorm (particularly about the items marked as (*))

Depends on https://github.com/apple/swift-syntax/pull/2623

There's also a corresponding PR in swift-driver, which is blocked by this one: https://github.com/apple/swift-driver/pull/1582

kabiroberai committed:
wip

kabiroberai committed:
wip

kabiroberai committed:
wip

kabiroberai committed:
revert swift-plugin-server

kabiroberai committed:
use jsc in-proc

kabiroberai committed:
Allow wasm in sandbox, don’t disable it

kabiroberai committed:
fix warning, prune version check

kabiroberai committed:
Factor out JSCWasmPlugin

kabiroberai committed:
the continuation just... works now?

kabiroberai committed:
more tweaks to JSCWasmPlugin

kabiroberai committed:
Create SwiftPluginServerSupport

kabiroberai commented on lib/AST/PluginLoader.cpp:
```diff
@@ -106,8 +106,19 @@ PluginLoader::getPluginMap() {
+        // "macro runner" executable
+        SmallString<128> runner;
+        // TODO: improve path resolution: we really want tools_dir
+        llvm::sys::path::append(runner, Ctx.SearchPathOpts.RuntimeResourcePath, "../../bin/swift-wasm-plugin-server");
```
This seems like it's not the best way to find the binary — is there a better way to locate it from the frontend or should we hand it down from the driver?

kabiroberai replied:
Decided to tentatively use the Driver approach.

kabiroberai commented on tools/swift-plugin-server/allow-jit.plist:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+<key>com.apple.security.cs.allow-jit</key>
```
Necessary in order for JSC to expose `globalThis.WebAssembly` — I don't think there's any security risk here because the same surface area is exposed via `/System/Library/Frameworks/JavaScriptCore.framework/Versions/Current/Helpers/jsc`.

kabiroberai committed:
JSCWasmPlugin: memoize, improve errors

kabiroberai commented:
paging @DougGregor @MaxDesiatov as stakeholders, would love to get your thoughts on this

kabiroberai committed:
steps towards x-plat

kabiroberai committed:
more modularity

kabiroberai commented:
also wondering how I should add end-to-end tests for this: I would include some wasm blob fixtures but I don't want to be mistaken for the Second Coming of Jia Tan

kabiroberai committed:
Add wasm-plugin-server-path frontend opt

kabiroberai commented:
Another question: how much should we generalize this right now? As in, should we have a specialized wasm executor embedded into `swift-wasm-plugin-server` or would it maybe make sense to ship a general `swift-wasm-runtime` that can be used for other things (such as Swift Package Plugins) as well? I'm leaning towards the former because I think we shouldn't prematurely abstract but there are pros to the other approach as well.

kabiroberai committed:
Start adding WasmKit support

kabiroberai committed:
Update WasmKitPlugin

kabiroberai committed:
safety, tweaks

kabiroberai commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -23,8 +32,32 @@ if (SWIFT_BUILD_SWIFT_SYNTAX)
+  )
+  if(SWIFT_HOST_VARIANT_SDK IN_LIST SWIFT_DARWIN_PLATFORMS)
+    add_custom_command(TARGET swift-wasm-plugin-server POST_BUILD COMMAND
+      codesign -fs - $<TARGET_FILE:swift-wasm-plugin-server> --entitlements ${CMAKE_CURRENT_SOURCE_DIR}/allow-jit.plist
```
not sure if this is sufficient: do we need a corresponding codesign invocation in build_script as well?

kabiroberai committed:
tweaks to JSC

kateinoigakukun reviewed (commented):
I'm not sure if adding JSC as a toolchain dependency is realistic, but generally looks good!

> how much should we generalize this right now? As in, should we have a specialized wasm executor embedded into swift-wasm-plugin-server or would it maybe make sense to ship a general swift-wasm-runtime that can be used for other things (such as Swift Package Plugins) as well?

I prefer the former too at this moment.

Also, it would be nice to have a custom section or exported function in the plugin .wasm file to retrieve the ABI version used to compile plugin executable for future ABI flexibility. We already have `getCapability` message for the protocol version but I think we need a separate interface for ABI because it must be determined before executing the plugin process.

kabiroberai commented:
> it would be nice to have a custom section or exported function
ah good call, adding another exported function for this purpose should be trivial. Though I'm wondering what the most future-proof way to do this would be; I'm not familiar with the Component Model internals yet but will we still be able to have an exported function if we adopt the Component Model for plugins in the future?

kateinoigakukun commented:
@kabiroberai Componentized .wasm file has a completely different binary format, so I don't think we can handle it at this moment. But custom section way would be friendly for other tools that don't have ability to execute wasm.

kabiroberai commented:
@kateinoigakukun makes sense, yeah I suppose we can always check the header to see if it's a Core Wasm v1 file or a component/future Wasm version. As LLVM doesn't directly map MC sections to Wasm sections I wasn't able to use `@_section` to declare the version but I got it working with [some inline asm](https://github.com/kabiroberai/Wacro/blob/18a568f2f8489a9c256493439305c172e45effc5/Sources/CWacroPluginRaw/section.c). Once we sort out the shape of the ABI I'm thinking I'll make a PR to swift-syntax to make `SwiftCompilerPlugin` itself export the `macro_*` functions (+ the custom section) under an `os(WASI)` check.

kabiroberai committed:
Proper WasmKit integration

kabiroberai committed:
tweaks

kabiroberai committed:
Check ABI version

kabiroberai committed:
move ABI check into WasmEnginePlugin

kabiroberai committed:
build fixes

kabiroberai committed:
Elide copies of wasm blob

kabiroberai committed:
fix duplicate libraries warning

kabiroberai committed:
add SWIFT_WASM_USE_JSC

kabiroberai commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -1,4 +1,16 @@
+  include(DarwinSDKs)
+  swift_get_host_triple(SWIFT_HOST_TRIPLE)
+
+  # override the remote SwiftSystem dependency in WasmKit
+  FetchContent_Declare(SwiftSystem SOURCE_DIR "${PROJECT_SOURCE_DIR}/../swift-system")
+
+  FetchContent_Declare(WasmKit SOURCE_DIR "${PROJECT_SOURCE_DIR}/../wasmkit")
+  FetchContent_MakeAvailable(WasmKit)
```
Should we be declaring `SWIFT_PATH_TO_SWIFT_SYSTEM_SOURCE` and `SWIFT_PATH_TO_WASMKIT_SOURCE` instead of hardcoding these paths? Curious why the `SWIFT_PATH_TO_*` variables are used despite the assumption that `swift` has those dirs as siblings courtesy of `update-checkout`.

kabiroberai committed:
Skip wasmkit CLI for now

kabiroberai committed:
Or actually build wasmkit-cli

kabiroberai committed:
use stdio for ipc

kabiroberai committed:
No need for memory witness

kabiroberai committed:
Byte-swap length if needed

kabiroberai committed:
Improve error handling

kabiroberai commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -23,8 +46,50 @@ if (SWIFT_BUILD_SWIFT_SYNTAX)
     Sources/CSwiftPluginServer/include
   )
+
+  set(SWIFT_WASM_ALLOW_JSC ON CACHE BOOL "Use JavaScriptCore Wasm runtime if possible" FORCE)
```
Wanted to leave this in for the initial review to show what the JSC runtime would look like, but I can remove this before we do a final pass since Doug mentioned that it's better to stick to a uniform implementation to begin with, for simplicity. I think it's definitely worth revisiting this in a follow-up because the JSC runtime is a lot faster than WasmKit when it can be used (500ms vs 1100ms in some cases I've tested.)

kabiroberai replied:
Split out the JSC stuff into a separate branch, happy to make a followup PR when we're ready for JSC: https://github.com/kabiroberai/swift/compare/kabir/wasm-plugins...kabir/jsc-wasm

kateinoigakukun commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -277,6 +277,11 @@ def Raccess_note_EQ : Joined<["-"], "Raccess-note=">,
   : Separate<["-"], "blocklist-file">, MetaVarName<"<path>">,
     HelpText<"The path to a blocklist configuration file">;
+
+def wasm_plugin_server_path : Separate<["-"], "wasm-plugin-server-path">,
```
Can we merge wasm plugin server into the main plugin server? Having different frontend options for each mode complicates build systems, so it would be nice to unify the plugin servers into single one and switch underlying plugin mechanisms based on plugin extension or file magic.

kabiroberai replied:
I have a few concerns about bringing wasm into the main plugin server, curious to hear what you folks think:

1. Security: merging the servers will mean that enabling JIT for Wasm would allow library-based plugins to themselves use JIT. This increases risk exposure a bit, though it might be mitigated by AMFI library validation if that's indeed applied to the plugin server.
2. Wasm plugins are currently invoked via `-load-plugin-executable Lib.wasm#Module`, with the server being an impl detail. Meanwhile bona fide server plugins are invoked with `-external-plugin-path Lib#Exec`, and notably A) this doesn't allow supplying an external module name, B) the client would have to locate the executable themselves, as afaik the Driver only does the locating for a pre-existing set of modules. Perhaps the solution to this would be to A) add a `-load-external-plugin-with-module Lib#Exec#Module` to the driver and B) Let the driver convert `-load-plugin-executable Lib.wasm#Module` to `-load-external-plugin-with-module Lib.wasm#<swift-plugin-server>#Module`? Would be nice to have the `-load-external-plugin-with-module` god-option anyway.

kateinoigakukun replied:
> 1. Security

Yes, once we will conclude that we cannot satisfy performance expectations without JIT, we need a separate process to minimize such risks, but let's start in a simpler way

> 2. God option

`-load-plugin <library path>:<server path>#<module name>,...` in the current tree seems good to me

DougGregor commented on tools/swift-plugin-server/Sources/swift-wasm-plugin-server/WasmKitEngine.swift:
```diff
@@ -0,0 +1,50 @@
+
+typealias DefaultWasmEngine = WasmKitEngine
+
+struct WasmKitEngine: WasmEngine {
+  private let module: Module
+  private let instance: ModuleInstance
+  private let runtime: Runtime
+
+  init(wasm: UnsafeByteBuffer, imports: WASIBridgeToHost) throws {
+    // we never call wasm.deallocator, effectively leaking the data,
+    // but that's intentional because plugins can't be "unloaded"
+    module = try parseWasm(bytes: Array(wasm.data))
+    runtime = Runtime(hostModules: imports.hostModules)
+    instance = try runtime.instantiate(module: module)
+  }
+
+  func customSections(named name: String) throws -> [ArraySlice<UInt8>] {
+    module.customSections.filter { $0.name == name }.map(\.bytes)
+  }
+
+  func invoke(_ method: String, _ args: [UInt32]) throws -> [UInt32] {
+    try runtime.invoke(instance, function: method, with: args.map(Value.i32)).map(\.i32)
+  }
```
This is shockingly simple. I love it

DougGregor reviewed (approved):
This is fantastic! My main comment is that I'd like to narrow down to WasmKit and link that support directly into swift-plugin-server.

kabiroberai commented:
appreciate the review @DougGregor! will be a bit busy next week (preparing my DeepDishSwift talk) but I'll respond to the review comments as soon as time permits; I like Yuta's suggestion re merging `swift-wasm-plugin-server` into `swift-plugin-server`, but I have a couple of practical (albeit surmountable) concerns. Also realized I should probably wait for #72999 to be merged and then rebase on top of that — cc @rintaro for viz.

kabiroberai committed:
Fix linkage

kabiroberai committed:
No C dep needed for wasm

kabiroberai commented on tools/swift-plugin-server/Package.swift:
```diff
@@ -1,38 +1,68 @@
+        "CSwiftPluginServer",
+        "SwiftPluginServerSupport",
+      ],
+      swiftSettings: [.interoperabilityMode(.Cxx)]
```
iirc this was required for swiftLLVMJSON but I was able to remove it now that we're using SwiftSyntax's JSON parser

kabiroberai commented on tools/swift-plugin-server/Sources/swift-wasm-plugin-server/swift-wasm-plugin-server.swift:
```diff
@@ -0,0 +1,129 @@
+
+@_spi(PluginMessage) import SwiftCompilerPluginMessageHandling
+@_spi(PluginMessage) import SwiftPluginServerSupport
+import Foundation
```
Hmm I managed to drop Foundation from the code here but I realized that WasmKit depends on it in a few places. Could we perhaps disable Wasm Plugin support when bootstrapping swift-foundation, if that's the concern?

Also worth mentioning: the primary load-bearing use case for Foundation here is using `NSData` with `mappedIfSafe`. It's easy to replace this with `System.FileDescriptor` but that doesn't support memory mapping afaik, so it'd contribute more to the dirty memory footprint. Any suggestions on other ways to do this? I'm working with the facts that 1. (non-embedded) Wasm binaries are currently 30mb+, and 2. we only need read access to them.

kabiroberai replied:
Update: working on dropping Foundation from WasmKit — https://github.com/swiftwasm/WasmKit/pull/91

kabiroberai commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -1,4 +1,17 @@
 if (SWIFT_BUILD_SWIFT_SYNTAX)
+  # hack: override macOS deployment target to 10.15.4
+  # since that's the minimum required by WasmKit
+  set(SWIFT_DARWIN_DEPLOYMENT_VERSION_OSX "10.15.4")
+  include(DarwinSDKs)
```
Also included this in https://github.com/swiftwasm/WasmKit/pull/91

kabiroberai committed:
Drop foundation (mostly); lower min os

kabiroberai committed:
More foundation pruning

kabiroberai committed:
interop mode cxx -> c

kabiroberai committed:
Target

kabiroberai committed:
tweaks

kabiroberai committed:
We don’t need WasmKitPluginError

kabiroberai committed:
Change wasm versioning approach

kabiroberai committed:
swift-wasm-plugin-server => swift-plugin-server

kabiroberai committed:
swift-plugin-server doesn’t need a product

kabiroberai committed:
We don’t need Dispatch here

kabiroberai committed:
Improve diagnostics

kabiroberai committed:
Remove JSC for now

kabiroberai committed:
Explicitly depend on System

kabiroberai committed:
Nevermind SwiftPM didn’t like that

kabiroberai committed:
Lower swift-tools-version

kabiroberai committed:
Fix WASI dep

kabiroberai committed:
Add missing header

kabiroberai committed:
Include position in diagnostics if possible

kabiroberai committed:
this comment is now stale

kabiroberai committed:
No wasm_plugin_server_path

kabiroberai committed:
Remove debug code

kabiroberai committed:
error_load_plugin diagnostic

kabiroberai commented on lib/AST/ModuleDependencies.cpp:
```diff
@@ -584,6 +584,14 @@ void SwiftDependencyTracker::addCommonSearchPathDeps(
+    case PluginSearchOption::Kind::LoadPlugin: {
+      auto &val = entry.get<PluginSearchOption::LoadPlugin>();
+      FS->status(val.LibraryPath);
+      // TODO: what else do we need?
```
not familiar with this part of the codebase — should I be recording `val.ServerPath` as well?

kateinoigakukun replied:
Yes, I think we need to record `.wasm` files under the search path

DougGregor commented on tools/swift-plugin-server/Sources/swift-plugin-server/WasmEngine.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+    try hostToPlugin.writeAll(json)
+
+    _ = try pumpFunction([])
```
I don't understand why we're passing in and returning unused `[UInt64]` instances to the pump function. Why pass anything if we aren't going to use it anywhere?

kabiroberai replied:
This is vestigial from when we were passing and returning JSON via function arguments instead of stdio; will remove.

DougGregor commented on tools/swift-plugin-server/Sources/swift-plugin-server/WasmMessageHandler.swift:
```diff
@@ -0,0 +1,114 @@
+@_spi(PluginMessage) import SwiftCompilerPluginMessageHandling
+
+/// A `PluginMessageHandler` that intercepts messages intended for Wasm plugins.
+final class WasmInterceptingMessageHandler<Base: PluginMessageHandler>: PluginMessageHandler {
```
This is effectively letting us chain message handlers, albeit in a specific way. It's fine as-is, but I could see us refactoring this later.

kabiroberai replied:
Agreed, I was tempted to generalize it but I realized that might be premature — think it's worth following the rule of three here.

DougGregor commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -1,10 +1,21 @@
 if (SWIFT_BUILD_SWIFT_SYNTAX)
+  # override the remote SwiftSystem and ArgParser dependencies in WasmKit
+  FetchContent_Declare(SwiftSystem SOURCE_DIR "${PROJECT_SOURCE_DIR}/../swift-system")
+  FetchContent_Declare(ArgumentParser SOURCE_DIR "${PROJECT_SOURCE_DIR}/../swift-argument-parser")
```
We might end up having to do something more bespoke here to get things building on all of the platforms, but let's try this to start!

DougGregor reviewed (commented):
I like the structure of this a lot: it makes the plugin server the one place we really have to deal with Wasm, and it's factored nicely. Do we have a mechanism by which we can write a basic test (akin to `test/Macros/macro_plugin_basic.swift`) to ensure that we keep it working?

kabiroberai commented on tools/swift-plugin-server/Sources/swift-plugin-server/WasmKitEngine.swift:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  func function(named name: String) throws -> WasmFunction? {
+    guard case let .function(function) = instance.exportInstances.first(where: { $0.name == name })?.value else {
```
It looks like there's a `ModuleInstance.exports: [String: ExternalValue]` but counter-intuitively it's a computed property (built from the `ModuleInstance.exportInstances` prop we're using here) so it'd probably be _slower_ than the O(n) pass we're doing, as it involves constructing a new dictionary each time. But we could cache it ourselves; I'll switch to doing that — thanks for the callout.

kabiroberai committed:
Feedback

kabiroberai committed:
nit

kabiroberai commented:
@DougGregor regarding tests, that's the last bit I want to figure out before marking this as ready for review. I'll see if I can do something bare-bones like `macro_plugin_basic.swift` but I'm not sure we can do even that without depending on WASI — and I don't know if we have a WASI SDK at test time, but I'll give it a shot. @kateinoigakukun do you have any ideas regarding adding tests of this sort?

Another option (which naturally lends itself to this approach) could be to add a pre-built Wasm executable fixture, though I'm unsure how we'd handle supply chain concerns, cf the recent xz backdoor.

kabiroberai committed:
Basic test

kabiroberai committed:
Parameterize

kabiroberai commented:
Update: I added a `wasm_sdk_root` parameter to lit.cfg such that the test suite now works but currently requires explicitly passing the sdk

```
lit.py -sv \
  -D wasm_sdk_root=$(echo ~/Library/org.swift.swiftpm/swift-sdks/swift-wasm-*-wasm32-unknown-wasi.artifactbundle/*-wasm32-unknown-wasi/wasm32-unknown-wasi) \
  ../build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/test-macosx-arm64/Macros/macro_plugin_wasm.swift
```

would appreciate suggestions on getting an SDK on CI

kabiroberai commented on test/Macros/Inputs/wasm_plugin.swift:
```diff
@@ -0,0 +1,210 @@
+import Foundation
```
can we use Foundation in tests at this level? Specifically, this would need to be a copy of Foundation built for Wasm, so layering concerns might apply here. We only need it for JSON parsing so maybe it's a better idea to use the JSON coders in SwiftSyntax — which are already guaranteed to support Wasm targets by virtue of this whole endeavor. I've confirmed that SwiftSyntax.JSON does indeed work in this environment by copying the `JSON/*` source files over (the only source change I made was `s/_SwiftSyntaxCShims/WASILibc/g`) but I assume we'll want to cross-compile SwiftSyntax with CMake as a more robust approach. I haven't attempted to cross-compile it with CMake yet — is that something that should work out of the box?

kateinoigakukun commented:
> I'm not sure we can do even that without depending on WASI — and I don't know if we have a WASI SDK at test time, but I'll give it a shot. @kateinoigakukun do you have any ideas regarding adding tests of this sort?

Sorry for being late. Currently, WASI Swift stdlib and runtime are built as a part of a separate build-script product from the main Swift build-script product, so we can't use Swift stdlib and runtime compiled for WASI while running `test/Macros` test suites.
Once we start building WASI stdlib as a part of the main Swift build-script product after [Evan's build system overhaul](https://github.com/etcwilde/swift/commits/wip-stdlib-build-overhaul/), the restriction will be relaxed, but it's not a feasible option right now.

Having said that we still have two options:

1. Given that what we want to test here is that Wasm macro program can properly communicate with plugin server but not that Wasm macro program *compiled from Swift* works fine, we can write a mock macro program in bare C instead of Swift. We can't use wasi-libc or libcxx here, so we need to call WASI functions (like `fd_write`) directly.
2. WASI stdlib is available at testing time of `WasmStdlib` build-script product, so we can run the test case at this step. Note that we build this product only in [WebAssembly CI job](https://ci.swift.org/job/oss-swift-pr-test-crosscompile-wasm-ubuntu-20_04/), so it's easy to be stale.

kabiroberai committed:
working test

kabiroberai committed:
simplify

kabiroberai committed:
error tests

kabiroberai commented:
@kateinoigakukun nice idea re declaring WASI functions directly! Was able to write a handful of unit tests for validating compiler/plugin-server behavior. My only concern is that we don't have any end-to-end tests now to ensure that macros built in _Swift_ work, but I suppose we could do those as a followup and scope them to the WebAssembly CI job. What do you think?

kateinoigakukun commented:
> I suppose we could do those as a followup and scope them to the WebAssembly CI job. What do you think?

@kabiroberai Thanks for adding a bunch of tests! I think it's reasonable to follow up e2e tests later.

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@swift-ci build toolchain

kabiroberai commented:
Ah I think CI is failing because we need to update WasmKit. @kateinoigakukun do you mind pushing a new tag? I'll bump the update-checkout pin after that.

kateinoigakukun commented:
@kabiroberai Ok, pushed a new tag https://github.com/swiftwasm/WasmKit/releases/tag/0.0.5

kateinoigakukun commented:
@swift-ci smoke test

kabiroberai commented:
@kateinoigakukun I believe the last test run failed because https://github.com/swiftlang/swift-syntax/pull/2623 hadn't been merged yet. That shouldn't be an issue now; mind running CI again please?

kateinoigakukun commented:
@swift-ci smoke test

kabiroberai commented:
@kateinoigakukun sorry you might need to run it again, just fixed a merge issue via e89fd6a 😶

I've asked Doug what I need to do to get CI perms myself — hopefully won't need to ask you soon enough 🤞

kateinoigakukun commented:
@swift-ci Please smoke test

No problem, ping me anytime :)

kateinoigakukun commented:
@swift-ci  smoke test

kateinoigakukun commented:
@swift-ci smoke test

kabiroberai committed:
no opaque types

kabiroberai commented:
Should have fixed the macOS build issue with 2b47724. Still trying to debug what went wrong on Linux though — seems that the `FetchContent_Declare(SwiftSystem ...)` is causing issues when we get to the lldb build. Presumably the same issue could affect macOS once it reaches that point.

```
...
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.10") 
CMake Error at cmake/modules/LLDBStandalone.cmake:12 (find_package):
  Found package configuration file:

    /home/build-user/build/buildbot_linux/swift-linux-x86_64/lib/cmake/swift/SwiftConfig.cmake

  but it set Swift_FOUND to FALSE so package "Swift" is considered to be NOT
  FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  SwiftSystem::SystemPackage

Call Stack (most recent call first):
  CMakeLists.txt:42 (include)


-- Configuring incomplete, errors occurred!
See also "/home/build-user/build/buildbot_linux/lldb-linux-x86_64/CMakeFiles/CMakeOutput.log".
```

kabiroberai committed:
tentative fix for lldb build

kabiroberai commented:
update: I think I figured it out — can't rely on `PROJECT_SOURCE_DIR` always being `/swift-project/swift`. @kateinoigakukun mind triggering CI again? ty 🙏

kateinoigakukun commented:
@swift-ci smoke test

kabiroberai committed:
don’t depend on ArgumentParser

kabiroberai commented:
I was able to repro this locally (added `--libcxx --lldb` to my local build-script invocation) and discovered that the root cause is that we're erroneously re-exporting WasmKit to `build/**/lib/cmake/swift/SwiftExports.cmake`, which was turning SwiftSystem into a transitive requirement for all dependents (such as lldb). The fix is to avoid adding WasmKit libs to `SWIFT_EXPORTS` — created https://github.com/swiftwasm/WasmKit/pull/101 for this.

kabiroberai commented:
updated wasmkit to point to my branch until we can tag 0.0.6, so CI should now work — hopefully we should get past the error this time 🤞

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun committed:
[test] Annotate macro tests that require the WebAssembly code generator

kateinoigakukun commented:
@swift-ci smoke test

kabiroberai commented:
nice, Linux and macOS tests are passing! Looks like Windows is failing because WasmKit's SystemExtras and WASI targets don't work there — @kateinoigakukun are there any major blockers to implementing those? If not, is there any way I can help move that along? Happy to contribute support for the necessary APIs, just want to make sure I'm not duplicating effort if you're already working on it.

kateinoigakukun commented:
@kabiroberai Actually, I have WIP change set [here](https://github.com/swiftwasm/WasmKit/pull/104) but it's still not working  at all. Given that most of the difficulties here come from the file system and we don't allow file system access at macro evaluation time, it might make sense to provide a subset of WASI syscall implementations excluding fs for Windows for now.
I don't have time to work on it at least this week, so if you are interested in, your contribution is welcome :)

DougGregor commented:
I like @kateinoigakukun 's idea of seeing whether we can slim down the WASI build to include those things that already work across the platforms and are needed by macros.

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun committed:
Add temporary workaround for Windows pipe support of swift-system

The support for `pipe` on Windows was added by https://github.com/apple/swift-system/commit/cfd3f9da906cf62efafd162d524df6334cc85da6
but it is not included in the latest swift-system release 1.3.1.
This commit adds a temporary inlined implementation of `pipe` API for
Windows until we update to the next release of swift-system.

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun committed:
[test] Remove usage of unused lit variables for building wasm plugins

`%c-flags` and `%exe-linker-flags` are derived from `CMAKE_C_FLAGS` and
`CMAKE_EXE_LINKER_FLAGS` respectively, which are not configured for Wasm
targets. And also those variables are not set in Windows build. So, this
patch removes the usage of those variables.

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
```
T:\5\tools\swift\test-windows-x86_64\Macros\Output\macro_plugin_wasm_guest_error.swift.tmp/test.swift:1:33: error: unexpected warning produced: external macro implementation type 'MacroDefinition.ConstMacro' could not be found for macro 'constInt()'; no such file or directory

@freestanding(expression) macro constInt() -> Int = #externalMacro(module: "MacroDefinition", type: "ConstMacro")
```

@kabiroberai  Hmm, something wrong is happening on Windows 🤔

kabiroberai commented:
thanks for all the Windows-related fixes @kateinoigakukun, looks like we're near the finish line! I wonder if the current failures are related to dos/unix path formats (backslash vs forward slash)?

ended up being busier than expected last week and might not have much time until EOW, but I'll try to configure a local testing setup on Windows by Saturday so that I can poke around. Wouldn't be surprised if it's a 1-2 line fix.

kateinoigakukun committed:
Change the first `-load-plugin` path separator from `:` to `#`

The ':' character can cause conflicts in Windows paths (e.g., C:\path\to\file).
The '#' character is less likely to appear in file paths and we already
use it for the second path separator in `-load-plugin` option

kateinoigakukun committed:
[test] Fix path segment separator of plugin path for Windows

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
@swift-ci smoke test

kateinoigakukun commented:
Okay, finally CI checks are all green!

etcwilde commented:
@swift-ci please Build Toolchain Linux Platform

MaxDesiatov commented:
@swift-ci build toolchain windows

DougGregor reviewed (commented):
This has ended up... almost straightforward in its integration. Very nice!

kabiroberai commented on tools/swift-plugin-server/CMakeLists.txt:
```diff
@@ -1,11 +1,24 @@
 if (SWIFT_BUILD_SWIFT_SYNTAX)
+  # override the remote SwiftSystem dependency in WasmKit
+  # TODO: create new variables instead of piggybacking off of SWIFT_PATH_TO_SWIFT_SYNTAX_SOURCE
+  file(TO_CMAKE_PATH "${SWIFT_PATH_TO_SWIFT_SYNTAX_SOURCE}" swift_syntax_path)
+  FetchContent_Declare(SwiftSystem SOURCE_DIR "${swift_syntax_path}/../swift-system")
+
+  set(WASMKIT_BUILD_CLI OFF)
+  FetchContent_Declare(WasmKit SOURCE_DIR "${swift_syntax_path}/../wasmkit")
+  FetchContent_MakeAvailable(WasmKit)
```
Indeed, do you think it's worth doing that right now? Or would it be preferable to merge as-is?

kateinoigakukun commented on test/Macros/macro_plugin_wasm_nostart.swift:
```diff
@@ -0,0 +1,33 @@
+}
+
+//--- MacroDefinition.c
+__attribute__((export_name("_tart")))
```
```suggestion
__attribute__((export_name("_start")))
```

MaxDesiatov commented:
@swift-ci test

kabiroberai commented:
macOS build failure seems related swift-testing (the [recent addition](https://forums.swift.org/t/an-update-about-swift-testing-in-the-swift-6-toolchain/74075) to the toolchain perhaps?)

```
[680/682] Compiling SwiftPMPackageTests runner.swift
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.derived/runner.swift:2:8: error: compiled module was created by an older version of the compiler; rebuild 'Testing' and try again: /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/testing/Testing.swiftmodule/x86_64-apple-macos.swiftmodule
  1 | #if canImport(Testing)
  2 | import Testing
    |        `- error: compiled module was created by an older version of the compiler; rebuild 'Testing' and try again: /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/testing/Testing.swiftmodule/x86_64-apple-macos.swiftmodule
  3 | #endif
  4 | 
error: fatalError
```

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@kateinoigakukun seems like WasmKit on `main` breaks something here:

```
1 | import Stringify
2 | print(#stringify(0))
3 | print(#stringify(1))
  |       `- error: failed to communicate with external macro implementation type 'StringifyMacros.StringifyMacro' to expand macro 'stringify'; callIndirectFunctionTypeMismatch(actual: WasmTypes.FunctionType(parameters: [WasmTypes.ValueType.i32, WasmTypes.ValueType.i32, WasmTypes.ValueType.i32, WasmTypes.ValueType.i32], results: []), expected: WasmTypes.FunctionType(parameters: [WasmTypes.ValueType.i32, WasmTypes.ValueType.i32, WasmTypes.ValueType.i32], results: [WasmTypes.ValueType.i32])) (from macro 'stringify')
```

Reproducible with a macOS toolchain built in https://ci.swift.org/job/swift-PR-toolchain-macos/1463/

kateinoigakukun commented:
@MaxDesiatov main branch of WasmKit is not stable yet as we replaced the VM design. Use tagged versions for now.

kateinoigakukun commented:
@swift-ci test

MaxDesiatov commented:
Heads up, this shouldn't be merged until it gets more approvals.

kabiroberai commented:
@MaxDesiatov is there anyone in particular worth pinging on Monday?

adrian-prantl commented:
Have you confirmed that these macros can be loaded by LLDB?
For example, by extending https://github.com/swiftlang/llvm-project/tree/next/lldb/test/API/lang/swift/macro ?

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
Update for WasmKit 0.1.0 API

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
> Have you confirmed that these macros can be loaded by LLDB? For example, by extending https://github.com/swiftlang/llvm-project/tree/next/lldb/test/API/lang/swift/macro ?

No, it can't be loaded since there's no support for Wasm in LLDB. We need https://github.com/llvm/llvm-project/pull/77949 or some analogous solution to land first.

adrian-prantl commented:
> > Have you confirmed that these macros can be loaded by LLDB? For example, by extending https://github.com/swiftlang/llvm-project/tree/next/lldb/test/API/lang/swift/macro ?
> 
> @adrian-prantl No, it can't be loaded since there's no support for Wasm in LLDB. We need [llvm/llvm-project#77949](https://github.com/llvm/llvm-project/pull/77949) or some analogous solution to land first.

Can you explain this dependency? Why does LLDB need to support Wasm in order to load a Wasm plugin into LLDB's embedded Swift compiler (that is used for expression evaluation and the Swift REPL)?

MaxDesiatov commented:
> Can you explain this dependency? Why does LLDB need to support Wasm in order to load a Wasm plugin into LLDB's embedded Swift compiler (that is used for expression evaluation and the Swift REPL)?

I see now, IIUC it doesn't need to debug Wasm macros directly. If it relies on the existing `swift-plugin-server` API, then nothing should change. Usage of Wasm under the hood is an implementation detail of `swift-plugin-server`.

MaxDesiatov committed:
Fix missing imports in `WasmKitEngine.swift`

MaxDesiatov commented:
@swift-ci build toolchain

kateinoigakukun commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
Bump stack size in `WasmKitEngine.swift` to avoid OOB

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci test webassembly

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@kabiroberai heads up, I resolved conflicts assuming that the option you added in this PR is superseded by `-load-resolved-plugin` introduced in https://github.com/swiftlang/swift/pull/76732

MaxDesiatov committed:
Remove outdated `case OPT_load_plugin` in `CompilerInvocation.cpp`

This case no longer exists after recent conflict resolution with `main`.

MaxDesiatov committed:
Remove outdated use of `LoadPlugin` from `swift-ide-test.cpp`

This symbol no longer exists after recent conflict resolution with `main`.

kabiroberai commented:
Thanks @MaxDesiatov! In fact I think we can remove the `-load-plugin` option from Options.td as well. Additionally I believe we'll need to fix up some of the tests to use `-load-resolved-plugin`, and also update the corresponding [swift-driver PR](https://github.com/swiftlang/swift-driver/pull/1582). I can try to get to this over the weekend.

On a related note, is there anything else I can do to help move things along here? Is the primary blocker atm related to the incubation of WasmKit?

MaxDesiatov committed:
Update lit tests to use `-load-resolved-plugin`

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
https://github.com/swiftlang/swift-syntax/pull/2915

@swift-ci test

MaxDesiatov committed:
Adopt new `shutDown` API introduced in `swift-syntax`

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
Add a trivial echo macro test to the package

MaxDesiatov committed:
Test `WasmEnginePlugin` APIs in `WasmEngineTests`

MaxDesiatov committed:
Make formatting consistent in `swift-plugin-server`

MaxDesiatov committed:
Create stdin/stdout pipes for plugins with `mkfifo`

This makes the API more portable, allowing file paths instead of file descriptors to be passed to Wasm engines.

MaxDesiatov committed:
Clean up temp Wasm fixture file in `WasmEngineTests`

MaxDesiatov committed:
Make `init` parameter names more explicit in `WasmEngine`

kateinoigakukun commented on tools/swift-plugin-server/Sources/swift-plugin-server/WasmEngine.swift:
```diff
@@ -10,42 +10,71 @@
+    self.stdoutPath = self.tempDirectory.appending("stdout")
+
+    let stdinResult = self.stdinPath.withCString {
+      mkfifo($0, S_IRUSR | S_IWUSR)
```
Why we need named pipe here? Wasn't `pipe(2)` enough for us?

MaxDesiatov replied:
As mentioned in the commit, this makes the API more general and easier to test.

MaxDesiatov replied:
I've cleaned it up in a follow-up commit (https://github.com/swiftlang/swift/pull/73031/commits/0da4f22081333a028eef181e70e1496c6ae89916)

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci test

MaxDesiatov committed:
Remove use of named pipes as engine-specific

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
Fix renaming in `tools/swift-plugin-server/CMakeLists.txt`

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
Fix Windows compatibility

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented on tools/swift-ide-test/swift-ide-test.cpp:
```diff
@@ -366,7 +366,6 @@ LoadPluginExecutable("load-plugin-executable",
                llvm::cl::desc("load plugin executable"),
                llvm::cl::cat(Category));
 
```
```suggestion


```

MaxDesiatov committed:
Update tools/swift-ide-test/swift-ide-test.cpp

MaxDesiatov commented:
@swift-ci smoke test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov committed:
swift-plugin-server: forward capabilities and handle version compat

MaxDesiatov commented on tools/swift-plugin-server/Package.swift:
```diff
@@ -13,13 +13,16 @@ let package = Package(
       name: "swift-plugin-server",
       dependencies: [
         .product(name: "_SwiftCompilerPluginMessageHandling", package: "swift-syntax"),
         .product(name: "_SwiftLibraryPluginProvider", package: "swift-syntax"),
```
I don't think this was fully handled in the PR previously, but I added some handling and testing for this in https://github.com/swiftlang/swift/pull/73031/commits/3799bf751b43afc846f79f23f27278c04349c6c0.

For backward compat (new compiler, old plugin), the host reads plugin's reported version and gates what it emits:
  - `Macros.swift` has `pluginProtocolVersion = macro.plugin.capability?.protocolVersion ?? 0`
  - `PluginHost.swift` has `kind = pluginProtocolVersion >= 8 ? .accessor : .declaration`.

Thus v8 protocol compiler talking to a v<8 plugin sends `.declaration` instead of `.accessor`.

For forward compat with old compiler, I verified that we have two independent mechanisms:

1. Plugin downgrades its response to the host's version. `hasExpandMacroResult: Bool { protocolVersion >= 5 }`; the plugin emits `expandMacroResult` only for host v≥5, otherwise the legacy `expandFreestandingMacroResult`/`expandAttachedMacroResult`.
2. Additive wire format:
    - New fields are optional with `= nil` defaults for `macroRole`, `lexicalContext`, `staticBuildConfiguration`
    - The decoder only fetches keys the type asks for and never iterates the JSON to reject
  unknown keys.
    - Explicit shims cover renames: `PluginMessageCompatibility.swift` decodes an old compiler's "freeStandingDeclaration" string back into `.declaration`.

MaxDesiatov commented:
@swift-ci smoke test

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci smoke test

MaxDesiatov commented:
@swift-ci smoke test macos

MaxDesiatov committed:
Post-rebase cleanup

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci smoke test
